docs/core/baseapp.md

```diff
@@ -7,167 +7,206 @@
 
-```
+![Types](./baseapp_state_types.png)
 
-### Main State
+### InitChain
 
-The main state is the canonical state of the application. It is initialized on [`InitChain`](#initchain) and updated on [`Commit`](#abci-commit) at the end of each block.
+During `InitChain`, the two volatile states, `checkState` and `deliverState` are set by cache-wrapping
+the root `CommitMultiStore`. Any subsequent reads and writes happen on cached versions of the `CommitMultiStore`.
 
-```
-+--------+                              +--------+
-|        |                              |        |
-|   S    +----------------------------> |   S'   |
-|        |   For each T in B: apply(T)  |        |
-+--------+                              +--------+
-```
+![InitChain](./baseapp_state-initchain.png)
+
+### CheckTx
 
-The main state is held by `baseapp` in a structure called the `CommitMultiStore`. This multi-store is used by developers to instantiate all the stores they need for each of their application's modules.
+During `CheckTx`, the `checkState`, which is based off of the last committed state from the root
+store, is used for any reads and writes. Here we only execute the `AnteHandler` and verify a router
+exists for every message in the transaction. Note, when we execute the `AnteHandler`, we cache-wrap
+the already cache-wrapped `checkState`. This has the side effect that if the `AnteHandler` fails,
+the state transitions won't be reflected in the `checkState` -- i.e. `checkState` is only updated on
+success.
+
+![CheckTx](./baseapp_state-checktx.png)
+
+### BeginBlock
 
-### Volatile States
+During `BeginBlock`, the `deliverState` is set for use in subsequent `DeliverTx` ABCI messages. The
+`deliverState` is based off of the last committed state from the root store and is cache-wrapped.
+Note, the `deliverState` is set to `nil` on [`Commit`](#commit-1).
 
-Volatile - or cached - states are used in between [`Commit`s](#commit) to manage temporary states. They are reset to the latest version of the main state after it is committed. There are two main volatile states:
+![CheckTx](./baseapp_state-begin_block.png)
 
-- `checkState`: This cached state is initialized during [`InitChain`](#initchain), updated during [`CheckTx`](#abci-checktx) when an unconfirmed transaction is received, and reset to the [main state](#main-state) on [`Commit`](#abci-commit).
-- `deliverState`: This cached state is initialized during [`BeginBlock`](#beginblock), updated during [`DeliverTx`](#abci-delivertx) when a transaction included in a block is processed, and reset to the [main state](#main-state) on [`Commit`](#abci-commit).
+### DeliverTx
+
+The state flow for `DeliverTx` is nearly identical to `CheckTx` except state transitions occur on
+the `deliverState` and messages in a transaction are executed. Similarly to `CheckTx`, state transitions
+occur on a doubly cache-wrapped state -- `deliverState`. Successful message execution results in
+writes being committed to `deliverState`.
```

Comment:
Perhaps worth noting that even if msg execution fails here, any state updates from successful antehandler get persisted.

Reply:
Updated.